File: packages/forge/lib/template.ts
import {
  fetchClient,
  clientToTemplate,
  buildClasspath,
  buildLaunch,
  mapLibraries,
} from '@opys/minecraft-vanilla';
import {
  type Artifact,
  type ValDefs,
  type Launch,
  sourceUrl,
  fetchWithRetry,
} from '@opys/core';
import type { Val, Valset } from '@opys/core';
import { mergeArgs, parseLibraries } from '@opys/mojang';
import {
  parseForgeRecipe,
  type ForgeRecipe,
  type LegacyLibrary,
} from './recipe';
import { resolveForgeVersion, type ForgeIndexEntry } from './resolver';
import {
  FORGE_WRAPPER_MAIN,
  stripModuleArgs,
  resolveForgeWrapperArtifact,
  buildForgeWrapperJvmArgs,
  type ForgeWrapperOptions,
} from '@opys/forgewrapper';

export type { ForgeWrapperOptions };

const DEFAULT_SOURCE = 'https://fuckforge.harmoniya.net';

export interface ForgeOptions {
  /**
   * Forge version. Accepts:
   *   - Bare MC version: `1.20.1` (resolves to the `best` Forge build)
   *   - Alias: `1.20.1-latest` | `1.20.1-recommended` | `1.20.1-best`
   *   - Full Forge build ID: `1.20.1-47.4.20`
   */
  version: string;
  /** fuckforge index base URL. Default: `https://fuckforge.harmoniya.net`. */
  source?: string;
  /** Override the bundled ForgeWrapper JAR (processor era only). */
  forgeWrapper?: ForgeWrapperOptions;
  /**
   * URL of the Mojang version manifest, when it is not Mojang's own — a
   * mirror, or a stand-in server under test. Passed straight to
   * `fetchClient`.
   */
  manifestBase?: string;
}

export interface ForgeTemplate {
  /** Vanilla MC + Forge runtime libraries, plus installer + ForgeWrapper for the processor era. */
  artifacts: Artifact[];
  vars: ValDefs;
  /** Assembled Launch — drop straight into `manifest.launch`. */
  launch: Launch;
  /** JVM args alone, for composition (e.g. interleaving authliberty's `-javaagent`). */
  jvmArgs: Valset;
  /** Main class wrapped as a `Val` so it spreads into a `Valset`. Raw string at `mainClass.value[0]`. */
  mainClass: Val;
  /** Game args alone, for composition. */
  gameArgs: Valset;
}

export async function resolveForge(
  options: ForgeOptions,
): Promise<ForgeTemplate> {
  const source = options.source ?? DEFAULT_SOURCE;
  const indexEntry = await resolveForgeVersion(options.version, source);

  if (!indexEntry.recipe) {
    throw new Error(
      `No recipe URL listed for Forge build '${indexEntry.forge}'`,
    );
  }

  const recipeRes = await fetchWithRetry(indexEntry.recipe);
  if (!recipeRes.ok) {
    throw new Error(
      `Failed to fetch Forge recipe from ${indexEntry.recipe}: ${recipeRes.statusText}`,
    );
  }
  const recipe = parseForgeRecipe(await recipeRes.json());

  if (recipe.kind === 'unsupported') {
    throw new Error(
      `Forge era '${recipe.type}' (Minecraft <1.7) is not yet supported by @opys/minecraft. ` +
        `Only 'legacy' (1.7–1.12) and 'processor' (1.13+) builds are supported.`,
    );
  }

  if (recipe.kind === 'legacy') {
    return buildLegacyTemplate(recipe, indexEntry, options);
  }
  return buildProcessorTemplate(recipe, indexEntry, options);
}

function legacyLibraryToArtifact(lib: LegacyLibrary): Artifact {
  // No integrity when the recipe has no sha1 — only the 1.6.x builds, whose
  // artifacts are gone from every maven, reach here without one.
  return {
    path: `\${library_directory}/${lib.path}`,
    source: sourceUrl(lib.url),
    rules: [],
    ...(lib.sha1 ? { integrity: { sha1: lib.sha1 } } : {}),
    ...(lib.size != null ? { size: lib.size } : {}),
  };
}

async function buildLegacyTemplate(
  recipe: Extract<ForgeRecipe, { kind: 'legacy' }>,
  indexEntry: ForgeIndexEntry,
  options: ForgeOptions,
): Promise<ForgeTemplate> {
  // The universal jar arrives as one of `recipe.libraries`, carrying its own
  // maven URL and sha1 — `indexEntry.files.universal` is not consulted.
  const { client } = await fetchClient(indexEntry.id, {
    manifestBase: options.manifestBase,
  });
  const mc = await clientToTemplate(client);

  const forgeLibArtifacts = recipe.libraries.map(legacyLibraryToArtifact);

  // Classpath: vanilla client.jar + Forge runtime libs + vanilla MC libs
  // (log4j, lwjgl, etc. — recipe.libraries doesn't repeat them since
  // legacy version.json inherits from vanilla). Forge entries go first so
  // any version conflict resolves in Forge's favor, matching what the
  // Forge installer's generated version.json does. Legacy era doesn't use
  // the JVM module system, so no -p filtering.
  const forgeCp = recipe.libraries.map((l) => ({
    rules: l.rules,
    artifactPath: `\${library_directory}/${l.path}`,
  }));
  const vanillaCp = client.libraries.map((l) => ({
    rules: l.rules,
    artifactPath: `\${library_directory}/${l.artifact.path}`,
  }));
  const classpathEntries = buildClasspath(
    [...forgeCp, ...vanillaCp],
    '${version_dir}/client.jar',
  );

  const vars: ValDefs = { ...mc.vars, classpath: classpathEntries };

  // Legacy `minecraftArguments` REPLACE vanilla's args entirely (per fuckforge
  // recipe semantics) — don't merge with `client.args`.
  const parts = buildLaunch(
    recipe.mainClass,
    recipe.args.game,
    recipe.args.jvm,
  );

  return {
    artifacts: [...mc.artifacts, ...forgeLibArtifacts],
    vars,
    ...parts,
  };
}

async function buildProcessorTemplate(
  recipe: Extract<ForgeRecipe, { kind: 'processor' }>,
  indexEntry: ForgeIndexEntry,
  options: ForgeOptions,
): Promise<ForgeTemplate> {
  const installerFile = indexEntry.files.installer;
  if (!installerFile) {
    throw new Error(
      `No installer file listed for Forge build '${indexEntry.forge}'`,
    );
  }
  if (!indexEntry.installProfile) {
    throw new Error(
      `No install_profile URL listed for Forge build '${indexEntry.forge}'`,
    );
  }

  // The recipe's `libraries[]` is the runtime classpath subset. The full set
  // of files Forge needs on disk lives in the installer's install_profile.json
  // (which fuckforge serves as a separate URL). It includes the fml*/forge:
  // universal libs that aren't on the launch classpath but are loaded by
  // BootstrapLauncher's JarJar/module-discovery via -DlibraryDirectory.
  const installProfileRes = await fetchWithRetry(indexEntry.installProfile);
  if (!installProfileRes.ok) {
    throw new Error(
      `Failed to fetch install_profile from ${indexEntry.installProfile}: ${installProfileRes.statusText}`,
    );
  }
  const installProfile = (await installProfileRes.json()) as {
    libraries?: unknown[];
  };
  // An install-profile library with no URL is not a download: the processors
  // produce it, and ForgeWrapper runs them. Forge ships these as `"url": ""`
  // — 1.13.2 through 1.16.5 each list their own universal jar that way — and
  // without this filter they became artifacts with an empty source URL.
  const installProfileLibs = parseLibraries(
    installProfile.libraries ?? [],
  ).filter((l) => l.artifact.url !== '');

  const { client } = await fetchClient(indexEntry.id, {
    manifestBase: options.manifestBase,
  });
  const mc = await clientToTemplate(client);

  // Forge's args APPEND to vanilla's args.
  const merged = mergeArgs(client.args, recipe.args);

  const installerPath = `\${library_directory}/net/minecraftforge/forge/${indexEntry.forge}/forge-${indexEntry.forge}-installer.jar`;
  const installerArtifact: Artifact = {
    path: installerPath,
    source: sourceUrl(installerFile.url),
    rules: [],
    integrity: { md5: installerFile.md5 },
  };

  const { artifact: forgeWrapperArtifact, path: forgeWrapperPath } =
    resolveForgeWrapperArtifact(options.forgeWrapper ?? {});

  // Download set: union of recipe runtime libs + install_profile libs.
  // deduplicateArtifacts (in core) collapses duplicates by path.
  const forgeLibArtifacts = [
    ...mapLibraries(recipe.libraries),
    ...mapLibraries(installProfileLibs),
  ];

  // Classpath: vanilla libs + recipe runtime libs + ForgeWrapper.
  // ForgeWrapper (PrismLauncher fork) handles module-path setup at runtime, so
  // all jars go on -cp; no -p filtering needed here.
  const cpLibs = [...client.libraries, ...recipe.libraries];
  const libPaths = cpLibs.map((l) => ({
    rules: l.rules,
    artifactPath: `\${library_directory}/${l.artifact.path}`,
  }));
  libPaths.push({ rules: [], artifactPath: forgeWrapperPath });

  const classpathEntries = buildClasspath(
    libPaths,
    '${version_dir}/client.jar',
  );

  const vars: ValDefs = { ...mc.vars, classpath: classpathEntries };

  const finalJvm = [
    ...buildForgeWrapperJvmArgs(installerPath),
    ...stripModuleArgs(merged.jvm),
  ];

  const parts = buildLaunch(FORGE_WRAPPER_MAIN, merged.game, finalJvm);

  return {
    artifacts: [
      ...mc.artifacts,
      ...forgeLibArtifacts,
      installerArtifact,
      forgeWrapperArtifact,
    ],
    vars,
    ...parts,
  };
}

File: crates/opys-forge/src/template.rs
//! The `forge` plugin: a published Forge version document, folded onto vanilla.

use opys_core::{Artifact, ConditionalVal, Launch, Val, ValDefs};
use opys_dev::http::get_json;
use opys_dev::{Contribution, LaunchFragment, PluginOutput};
use opys_minecraft_vanilla::{fetch_client, patch_to_template, resolve_client_template};
use opys_mojang::VersionPatch;
use serde::{Deserialize, Serialize};

use crate::error::ForgeError;
use crate::index::{resolve_forge_version, DEFAULT_FORGE_INDEX};

/// The name this plugin claims in the plugin map.
pub const PLUGIN_NAME: &str = "forge";

/// What to resolve.
#[derive(Debug, Clone, Default, PartialEq, Eq, Serialize, Deserialize)]
#[serde(default, rename_all = "camelCase")]
pub struct ForgeOptions {
    /// A Minecraft version (`1.20.1`), an alias (`1.20.1-recommended`), or a
    /// full Forge build id (`1.20.1-47.4.10`).
    pub version: String,
    /// Document index base URL. `None` is [`DEFAULT_FORGE_INDEX`].
    pub source: Option<String>,
    /// Overrides the canonical Mojang version-manifest URL — a mirror, or the
    /// seam the tests point at a loopback server.
    pub manifest_base: Option<String>,
}

/// Everything a Forge build plus its vanilla base contributes to a manifest.
#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct ForgeTemplate {
    /// Vanilla artifacts followed by Forge's own libraries.
    pub artifacts: Vec<Artifact>,
    pub vars: ValDefs,
    /// Per-OS classpath arms (also baked into `vars.classpath`), exposed so a
    /// plugin stacked on top of Forge can rebuild it with its own libraries.
    pub classpath: Vec<ConditionalVal>,
    /// Assembled launch — drop straight into `manifest.launch`.
    pub launch: Launch,
    /// JVM args alone, for composition (e.g. interleaving an auth `-javaagent`).
    pub jvm_args: Vec<Val>,
    /// Main class wrapped as a `Val` so it spreads into a `Valset`.
    pub main_class: Val,
    /// Game args alone, for composition.
    pub game_args: Vec<Val>,
}

/// Fetch and parse one build's version document.
pub fn fetch_document(url: &str) -> Result<VersionPatch, ForgeError> {
    Ok(get_json(url, &[])?)
}

/// Resolve Forge: pick the build, read its document, fetch the vanilla version
/// it inherits from, then fold the two together.
///
/// There is no era to branch on. Every Forge build — the 1.1 jar mods, the
/// 1.7-era LaunchWrapper releases, the 1.13+ processor installs — is published
/// as the same kind of document, and the differences between them are already
/// spelled out inside it: which libraries to add, which class to launch, which
/// properties ForgeWrapper needs. That is the whole reason the documents
/// exist, and it is what makes this the same shape as `opys-fabric`.
pub fn resolve_forge(options: &ForgeOptions) -> Result<ForgeTemplate, ForgeError> {
    let source = options.source.as_deref().unwrap_or(DEFAULT_FORGE_INDEX);
    let release = resolve_forge_version(&options.version, source)?;
    let patch = fetch_document(&release.document_url)?;

    let (_, client) = fetch_client(Some(&patch.inherits_from), options.manifest_base.as_deref())?;
    let vanilla = resolve_client_template(&client)?;
    let folded = patch_to_template(&patch, &client, &vanilla)?;

    Ok(ForgeTemplate {
        artifacts: folded.artifacts,
        vars: folded.vars,
        classpath: folded.classpath,
        launch: folded.launch,
        jvm_args: folded.jvm_args,
        main_class: folded.main_class,
        game_args: folded.game_args,
    })
}

/// A resolved template as a ready-to-merge contribution.
pub fn build_forge(options: &ForgeOptions) -> Result<PluginOutput, ForgeError> {
    let template = resolve_forge(options)?;
    Ok(PluginOutput {
        name: PLUGIN_NAME.to_owned(),
        contribution: Contribution {
            artifacts: template.artifacts,
            vars: template.vars,
            launch: [
                (
                    "command".to_owned(),
                    LaunchFragment::Text(template.launch.command.clone()),
                ),
                (
                    "jvmArgs".to_owned(),
                    LaunchFragment::Many(template.jvm_args),
                ),
                (
                    "mainClass".to_owned(),
                    LaunchFragment::One(template.main_class),
                ),
                (
                    "gameArgs".to_owned(),
                    LaunchFragment::Many(template.game_args),
                ),
            ]
            .into_iter()
            .collect(),
            envs: Default::default(),
        },
    })
}
